File: poc-csharp/RpsPoc/Ui/ArenaPanel.cs
using Avalonia;
using Avalonia.Controls;

using RpsPoc.Domain;

namespace RpsPoc.Ui;

/// <summary>
/// RF11 / UC05: o espaco se redivide sozinho conforme o numero de matches ativos.
/// 1 -> tela cheia; 2 -> 50/50; N -> grade quadrada mais proxima, celulas iguais.
/// </summary>
public sealed class ArenaPanel : Grid
{
    private readonly Dictionary<int, MatchView> _views = new();

    public ArenaPanel()
    {
        Margin = new Thickness(10);
    }

    public void Add(Match match, int maxScore)
    {
        _views[match.Id] = new MatchView(match, maxScore);
        Relayout(maxScore);
    }

    public void Remove(Match match, int maxScore)
    {
        _views.Remove(match.Id);
        Relayout(maxScore);
    }

    /// <summary>Repinta todos (o score de alguem pode ter mudado).</summary>
    public void RefreshAll(int maxScore)
    {
        foreach (var view in _views.Values)
        {
            view.Refresh(maxScore);
        }
    }

    public MatchView? View(Match match) => _views.GetValueOrDefault(match.Id);

    private void Relayout(int maxScore)
    {
        Children.Clear();
        ColumnDefinitions.Clear();
        RowDefinitions.Clear();

        var n = _views.Count;
        if (n == 0)
        {
            return;
        }

        var cols = (int)Math.Ceiling(Math.Sqrt(n));
        var rows = (int)Math.Ceiling((double)n / cols);

        for (var c = 0; c < cols; c++)
        {
            ColumnDefinitions.Add(new ColumnDefinition(new GridLength(1, GridUnitType.Star)));
        }

        for (var r = 0; r < rows; r++)
        {
            RowDefinitions.Add(new RowDefinition(new GridLength(1, GridUnitType.Star)));
        }

        var i = 0;
        foreach (var view in _views.Values)
        {
            view.Refresh(maxScore);
            SetColumn(view, i % cols);
            SetRow(view, i / cols);
            Children.Add(view);
            i++;
        }
    }
}

File: src/main/java/rps/ui/ArenaPanel.java
package rps.ui;

import java.util.LinkedHashMap;
import java.util.Map;

import javafx.geometry.Insets;
import javafx.scene.layout.Background;
import javafx.scene.layout.BackgroundFill;
import javafx.scene.layout.ColumnConstraints;
import javafx.scene.layout.CornerRadii;
import javafx.scene.layout.GridPane;
import javafx.scene.layout.Priority;
import javafx.scene.layout.RowConstraints;
import javafx.scene.paint.Color;

import rps.domain.Match;

/**
 * RF11 / UC05: o espaco se redivide sozinho conforme o numero de matches ativos.
 * 1 -> tela cheia; 2 -> 50/50; N -> grade quadrada mais proxima, celulas iguais.
 */
public final class ArenaPanel extends GridPane {

    private final Map<Integer, MatchView> views = new LinkedHashMap<>();

    /** Fixa no start (ADR-012); o default so evita NPE antes do primeiro torneio. */
    private ColorScale scale = ColorScale.forPlayers(2);

    public ArenaPanel() {
        setHgap(8);
        setVgap(8);
        setPadding(new Insets(10));
        setBackground(new Background(new BackgroundFill(Color.web("#0e1116"), CornerRadii.EMPTY, Insets.EMPTY)));
    }

    public void scale(ColorScale scale) {
        this.scale = scale;
    }

    public void add(Match match) {
        views.put(match.id(), new MatchView(match, scale));
        relayout();
    }

    public void remove(Match match) {
        views.remove(match.id());
        relayout();
    }

    /** Repinta todos (score de alguem pode ter mudado). */
    public void refreshAll() {
        views.values().forEach(MatchView::refresh);
    }

    public MatchView view(Match match) {
        return views.get(match.id());
    }

    private void relayout() {
        getChildren().clear();
        getColumnConstraints().clear();
        getRowConstraints().clear();

        int n = views.size();
        if (n == 0) {
            return;
        }
        int cols = (int) Math.ceil(Math.sqrt(n));
        int rows = (int) Math.ceil((double) n / cols);

        for (int c = 0; c < cols; c++) {
            ColumnConstraints cc = new ColumnConstraints();
            cc.setPercentWidth(100.0 / cols);
            cc.setHgrow(Priority.ALWAYS);
            getColumnConstraints().add(cc);
        }
        for (int r = 0; r < rows; r++) {
            RowConstraints rc = new RowConstraints();
            rc.setPercentHeight(100.0 / rows);
            rc.setVgrow(Priority.ALWAYS);
            getRowConstraints().add(rc);
        }

        int i = 0;
        for (MatchView view : views.values()) {
            view.refresh();
            view.setMaxWidth(Double.MAX_VALUE);
            view.setMaxHeight(Double.MAX_VALUE);
            GridPane.setHgrow(view, Priority.ALWAYS);
            GridPane.setVgrow(view, Priority.ALWAYS);
            add(view, i % cols, i / cols);
            i++;
        }
    }
}
